src/Symfony/Component/ObjectMapper/ObjectMapper.php

```diff
@@ -0,0 +1,282 @@
+
+            if (!\is_object($mappedTarget)) {
+                throw new MappingTransformException('Can not map to a non-object.');
+            }
+        }
+
+        if (!is_a($mappedTarget, $targetRefl->getName(), false)) {
+            throw new MappingException(\sprintf('Expected the mapped object to be an instance of "%s".', $mappingToObject ? $mappedTarget::class : $mappedTarget));
+        }
+
+        $objectMap[$source] = $mappedTarget;
+        $ctorArguments = [];
+        $constructor = $targetRefl->getConstructor();
+        foreach ($constructor?->getParameters() ?? [] as $parameter) {
+            $parameterName = $parameter->getName();
+            if (!$targetRefl->hasProperty($parameterName)) {
+                continue;
+            }
+
+            $property = $targetRefl->getProperty($parameterName);
+
+            // The mapped class was probably instantiated in a transform we can't write a readonly property
+            if ($property->isReadOnly() && ($property->isInitialized($mappedTarget) && $property->getValue($mappedTarget))) {
+                continue;
+            }
+
+            $ctorArguments[$parameterName] = $parameter->isDefaultValueAvailable() ? $parameter->getDefaultValue() : null;
+        }
+
+        try {
+            $refl = new \ReflectionClass($source);
+        } catch (\ReflectionException $e) {
+            throw new ReflectionException($e->getMessage(), $e->getCode(), $e);
+        }
+
+        $mapToProperties = [];
+        foreach ($refl->getProperties() as $property) {
+            if ($property->isStatic()) {
+                continue;
+            }
+
+            $propertyName = $property->getName();
+            $mappings = $this->metadataFactory->create($source, $propertyName);
+            foreach ($mappings as $mapping) {
+                if (($if = $mapping->if) && ($fn = $this->getCallable($if)) && !$this->call($fn, null, $source)) {
+                    continue;
+                }
+
+                $targetPropertyName = $mapping->target ?? $propertyName;
+                if (!$targetRefl->hasProperty($targetPropertyName)) {
+                    continue;
+                }
+
+                $value = $this->getSourceValue($source, $mappedTarget, $propertyName, $objectMap, $mapping);
+                $this->storeValue($targetPropertyName, $mapToProperties, $ctorArguments, $value);
+            }
+
+            if (!$mappings && $targetRefl->hasProperty($propertyName)) {
+                $value = $this->getSourceValue($source, $mappedTarget, $propertyName, $objectMap);
+                $this->storeValue($propertyName, $mapToProperties, $ctorArguments, $value);
+            }
+        }
+
+        if (!$mappingToObject && $ctorArguments) {
+            try {
+                $constructor?->invokeArgs($mappedTarget, $ctorArguments);
+            } catch (\ReflectionException $e) {
+                throw new ReflectionException($e->getMessage(), $e->getCode(), $e);
```

Comment:
Could `map()` just throw `MappingException`s that wrap the original exception when there is one?